File: PracticaBloqueos/EjemploLock.java

import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class EjemploLock {
    private final Lock lock = new ReentrantLock();  // Creamos un ReentrantLock
    private int contador = 0;

    public static void main(String[] args) {
        EjemploLock ejemplo = new EjemploLock();

        // Crear 3 hilos que incrementarán el contador
        for (int i = 1; i <= 3; i++) {
            new Thread(() -> ejemplo.incrementar()).start();
        }
    }

    public void incrementar() {
        lock.lock();  // Adquirimos el lock antes de modificar el recurso compartido
        try {
            contador++;
            System.out.println(Thread.currentThread().getName() + " ha incrementado el contador a: " + contador);
        } finally {
            lock.unlock();  // Aseguramos liberar el lock al final
        }
    }
}

File: PracticaBloqueos/EjemploLock.py
import threading

class EjemploLock:
    def __init__(self):
        self.lock = threading.Lock()  # Creamos un Lock
        self.contador = 0

    def incrementar(self):
        with self.lock:  # Adquirimos el lock antes de modificar el recurso compartido
            self.contador += 1
            print(f"{threading.current_thread().name} ha incrementado el contador a: {self.contador}")

def main():
    ejemplo = EjemploLock()

    # Crear 3 hilos que incrementarán el contador
    threads = []
    for i in range(1, 4):
        thread = threading.Thread(target=ejemplo.incrementar, name=f"Hilo {i}")
        thread.start()
        threads.append(thread)

    # Esperar a que todos los hilos terminen
    for thread in threads:
        thread.join()

if __name__ == "__main__":
    main()
